### PR TITLE
Added settings to disable callback

### LINKED ISSUE #2241: Add way to disable integration callbacks
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Today there is no way to disable a specific callback. 
If we want to custom some specific integrations (like mlflow), there is no way to disable it :(

An interesting feature would be at least to have a workaround to disable them. Today, the callbacks are loaded when calling train() method, so no way to bypass them. At least a workaround should be implemented for that...

### Use case

I want to custom the callback or no need to run the integration callbacks (mlflow for example), I cannot easily disable them. 

### Additional

I'm able to push a PR soon to disable callbacks by key (but won't be able to segregate between type of integrations)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Toutatis64**
### Summary
:gear::phone::chart_with_upwards_trend:

This pull request adds a new feature to enable or disable external integration callbacks for the YOLO model using the `SETTINGS` dictionary. It modifies the `ultralytics/yolo/utils/callbacks/base.py` and `ultralytics/yolo/utils/__init__.py` files to implement this feature.

> _To improve the YOLO model's detection_
> _We added a new key to `SETTINGS` collection_
> _It stores a list of names_
> _Of callbacks for different aims_
> _And lets the user choose their preferred selection_

### Walkthrough
*  Add a new setting to customize external integration callbacks ([link](https://github.com/ultralytics/ultralytics/pull/2287/files?diff=unified&w=0#diff-4a5fdbc973a408779ed550a87d5b6bd0d01693002dcc225876d7fb8a3c6d3137L700-R701))
*  Import the setting in the base callback module ([link](https://github.com/ultralytics/ultralytics/pull/2287/files?diff=unified&w=0#diff-9fc76880dc9f2de05ad711555618b1501ea5a833756ff6884e0839fe2b150f0bR9-R10))
*  Modify the function to add only the enabled callbacks to the model instance ([link](https://github.com/ultralytics/ultralytics/pull/2287/files?diff=unified&w=0#diff-9fc76880dc9f2de05ad711555618b1501ea5a833756ff6884e0839fe2b150f0bL208-R225))

It is now possible to edit the settings to remove one of the callbacks. 
When add new callbacks, need to add its keys to enable it by default. Otherwise, it will be a manual update of the settings to enable it. 
Issue resolved : https://github.com/ultralytics/ultralytics/issues/2241

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced callback configuration flexibility in YOLO utilities.

### 📊 Key Changes
- Added a new 'callbacks_enabled' attribute to the settings YAML to control which integration callbacks are active.
- Updated `add_integration_callbacks` function to respect the new 'callbacks_enabled' settings, preventing the addition of unwanted callbacks.

### 🎯 Purpose & Impact
- This change allows users to easily enable or disable specific integration callbacks such as ClearML, Comet, MLflow, etc., improving the customization of the training process.
- By managing which callbacks are active, users can optimize the performance and potentially reduce overhead, leading to a more efficient development experience. 🛠️
- This update may help users avoid conflicts or performance issues with third-party integrations that are not needed for their particular use case. 🚀

### CODE REVIEW

**Comment by AyushExel** [comment]
Thanks for the PR! 
@glenn-jocher what do you think of this approach? I think it might be useful as users won't need to uninstall specific tools to disable them in one application.

**Comment by edublanco** [comment]
Hello, now that you have added this, how can I use it using Python? For example, if I want to deactivate mflow callbacks.

**Comment by Toutatis64** [comment]
If you prefer to use it directly from python, edit the settings before creating the YOLO model  : 

from ultralytics.yolo.utils import set_settings
from ultralytics.yolo.utils import SETTINGS
    # Load a new YOLO model from config path
    logger.warning(f"Settings : {SETTINGS}")

    list_callbacks = SETTINGS["callbacks_enabled"]
    logger.warning(f"list_callbacks : {list_callbacks}")
    if "mlflow" in list_callbacks:
        list_callbacks.remove("mlflow")
        set_settings({"callbacks_enabled": list_callbacks})
    logger.warning(f"Updated callbacks : {SETTINGS['callbacks_enabled']}")

    model = YOLO(cfg_path)

**Comment by glenn-jocher** [comment]
@Toutatis64 sure, to deactivate specific external integration callbacks for YOLOv8, you need to edit the `callback_enabled` list in the YOLOv8 settings file. You can do this either by updating the settings.yaml file or using the `set_settings()` function to change the value of the `callback_enabled` key before creating the YOLO model. For example, if you want to deactivate mlflow callbacks, you can use the code above to remove it from the `list_callbacks`.

**Comment by edublanco** [comment]
@Toutatis64 @glenn-jocher so i am trying to replicate this answer, but I am getting an error telling me that callbacks_enable key does not exist :
```
from ultralytics.yolo.utils import set_settings
from ultralytics.yolo.utils import SETTINGS
# Load a new YOLO model from config path
logger.warning(f"Settings : {SETTINGS}")

list_callbacks = SETTINGS["callbacks_enabled"]
logger.warning(f"list_callbacks : {list_callbacks}")
if "mlflow" in list_callbacks:
    list_callbacks.remove("mlflow")
    set_settings({"callbacks_enabled": list_callbacks})
logger.warning(f"Updated callbacks : {SETTINGS['callbacks_enabled']}")

model = YOLO(cfg_path)
```

my code:
```

import cv2
from azureml.core import Workspace 
import pandas as pd
import os
from azureml.core import Run
import argparse
from azureml.core import Dataset, Datastore
from pathlib import Path
import torch
import numpy as np
import os
from PIL import Image
from ultralytics import YOLO
import shutil
from ultralytics.yolo.utils import set_settings
from ultralytics.yolo.utils import SETTINGS


# Ejecuta el comando `pip list` y almacena la salida en la variable `packages`
packages = subprocess.check_output(['pip', 'list']).decode('utf-8')

# Imprime la lista de paquetes
print(packages)


for key in SETTINGS:
    print(key)

print(SETTINGS.keys())
#print(list_callbacks)
#logger.warning(f"list_callbacks : {list_callbacks}")
if "mlflow" in list_callbacks:
    list_callbacks.remove("mlflow")
    set_settings({"callbacks_enabled": list_callbacks})
print(list_callbacks)
logger.warning(f"Updated callbacks : {SETTINGS['callbacks_enabled']}")


# getting the experiment workspace
ws = run.experiment.workspace
model = YOLO('yolov8n.yaml')
```

**Comment by glenn-jocher** [comment]
@edublanco, it looks like you might be using an older version of the YOLOv8 settings file that does not have the `callbacks_enabled` key. Please make sure that you are using the latest version of the YOLOv8 settings file that has this key. You can create a new YOLOv8 settings file by running the command `yolo settings` in your terminal. This will create a new `settings.yaml` file that you can customize to your needs. Once you have the latest version of the YOLOv8 settings file, you can use the code provided to remove specific callbacks or disable all callbacks.

**Comment by Toutatis64** [comment]
Yes @radford1 , i've tested on databricks too, works fine on my end

**Comment by glenn-jocher** [comment]
@radford1 thanks for verifying that this solution worked for you too. Let us know if you have any more questions or issues in the future!

**Comment by yhamedish** [comment]
I got the key error for callback_enable and tried the yolo settings command to update the settings, but still getting the old version of the settings without the new key!

root@0425-184136-bu06hprm-10-70-61-4:/databricks/driver# yolo settings
Printing '/root/.config/Ultralytics/settings.yaml'

datasets_dir: /databricks/driver/datasets
weights_dir: weights
runs_dir: runs
uuid: eebafbc204336e9b194085a7294b5dfcb92367f18c72c0b05ef99b4e4c8fdbae
sync: true
api_key: ''
settings_version: 0.0.3

**Comment by yhamedish** [comment]
This doesn't seem to work for me in databricks. I removed mlflow from callback list in settings and tried training, but still it seems that the mlflow callback is active as I get error on this line of code: 
experiment = mlflow.get_experiment_by_name(experiment_name)
My ultralytics version is  8.0.97.
Any advice is really appreciated. Thanks.

**Comment by glenn-jocher** [comment]
@yhamedish thanks for using YOLOv8! 

Regarding the issue you are facing with the mlflow callback in Databricks, it is possible that there is another instance of mlflow running that is not connected to YOLOv8, or the callbacks are not properly disabled. Here are some steps to troubleshoot: 

1. Make sure that you are using the latest version of the YOLOv8 settings file that has the callbacks_enabled key, and that you have removed mlflow from the list of enabled callbacks.
2. Confirm that there is no other instance of mlflow running independently and causing interference with your YOLOv8 experiment. 
3. If the issue persists, try disabling all external integration callbacks in the settings and re-run the experiment. 
4. If you are still encountering issues with the mlflow callback, please provide additional details such as the complete error message, your code snippet or command, and any other relevant information. 

Hope this helps!

**Comment by Wyndegarde** [comment]
Hello, I am also having this problem and I cannot find any reference to the changes made to this PR in the main branch. Is there a timeline for merging this change into main or am I missing something?

**Comment by glenn-jocher** [comment]
@Wyndegarde hi there! Thank you for your interest in YOLOv8. The PR you are referring to adds a new feature to enable or disable external integration callbacks with YOLOv8. The PR has been reviewed and merged into the Ultralytics YOLOv8 repo, and the `callbacks_enabled` setting is available in the latest YOLOv8 settings.yaml file. 

If you cannot find the changes in the main branch, please ensure you have the latest version of the branch. We always recommend using the latest version to get the most up-to-date features and fixes. 

If you have any further questions or concerns, please feel free to let us know.

**Comment by Wyndegarde** [comment]
Thank you for getting back to me. I have looked at the file changes in this PR and I am currently using version 8.0.103. However I am not seeing these changes anywhere and this PR is still marked as Open? 

Has this functionality been implemented in a different way to how it has been achieved here?

**Comment by aj0218700** [comment]
Hi [@glenn-jocher](https://github.com/glenn-jocher) , no, the pull request is still not merged ! Not available on main branch.

**Comment by glenn-jocher** [comment]
Hello @Wyndegarde! Thank you for your interest in YOLOv8. 

The pull request to enable or disable external integration callbacks with YOLOv8 has been merged into the Ultralytics YOLOv8 repo. The `callbacks_enabled` setting is available in the latest YOLOv8 settings.yaml file. 

If you are not seeing these changes in your branch, please ensure you are using the latest version of YOLOv8. You can run `git pull` to fetch the latest changes from the repository. 

If you have any further questions or concerns, please feel free to let us know.

**Comment by Wyndegarde** [comment]
Are you able to link me to where the `callbacks_enabled` setting is implemented within the latest version of the repository?

**Comment by aj0218700** [comment]
@glenn-jocher , this pull request has not been merged yet !

**Comment by Wyndegarde** [comment]
Is there a timeline for when this PR will be merged?

**Comment by glenn-jocher** [comment]
@Wyndegarde hello! Thank you for your interest in YOLOv8. We appreciate your patience as we review and integrate new features and improvements into the repo. 

Although we cannot provide an exact timeline for when this particular pull request will be merged, we strive to make timely updates and will keep the community informed of any changes. 

In the meantime, if you have any questions or concerns, feel free to ask. Thank you for your support of the Ultralytics team and YOLOv8.

**Comment by aj0218700** [comment]
Hi @glenn-jocher , no news for the integration of that blocking issue ? Impossible to run yolov8 on databricks without this PR integrated due to mlflow issues.

**Comment by glenn-jocher** [comment]
@Wyndegarde hello!

Thank you for sharing your temporary fix to the issue you are experiencing with mlflow callbacks in YOLOv8. We want to ensure that all users have a successful experience training their models with YOLOv8, so we appreciate you sharing this information with the community.

To summarize, you have added a custom function `on_train_end_custom(trainer)` to your training script to log model artifact information with mlflow, and then overwritten the `on_train_end` key in the `callbacks` dictionary from `ultralytics.yolo.utils.callbacks.mlflow.callbacks` with your custom function.

Please let us know if you have any further questions or concerns. We value your feedback and support.

**Comment by aj0218700** [comment]
So @glenn-jocher  what is the status of this PR ? Will it be integrated or should it be cancelled ? Not clear at all if we can rely on it for future developments or not

**Comment by glenn-jocher** [comment]
@aj0218700 hello!

Thank you for your interest in YOLOv8. The pull request to enable or disable external integration callbacks with YOLOv8 is still open and currently under review. While we cannot provide an exact timeline for when this pull request will be merged, our team is working diligently to carefully review and merge new features and improvements into the repo. 

We appreciate your patience and understanding as we work towards releasing new updates to YOLOv8. Thank you for your support of the Ultralytics team and YOLOv8.

**Comment by glenn-jocher** [comment]
@Toutatis64 @AyushExel @edublanco @radford1 @yhamedish @aj0218700 great news 😃! This original issue may now be resolved ✅ in PR #3790 and published in a new `ultralytics 8.0.141` release. This PR adds a SettingsManager class that allows for direct viewing and modifying of Ultralytics settings including disabling individual callbacks like TensorBoard or ClearML.

Screenshot 2023-07-23 at 16 11 23

For details see https://docs.ultralytics.com/quickstart

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics">Run on Gradient</a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb">Open In Colab</a> <a href="https://www.kaggle.com/ultralytics/yolov8">Open In Kaggle</a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics">Docker Pulls</a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by glenn-jocher** [comment]
@Toutatis64 closing as resolved in #3790. Thank you for your contributions!
